### PR TITLE
Specify conditional action dependencies (3.14)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch updates the Netsuke design and roadmap to make the conditional action model explicit before implementation work continues. It records what the current code supports, clarifies how dependency fields should lower into IR and Ninja, removes an unimplemented compiler dependency import example from the user guide, and tracks the replacement design as planned `deps_from` work.

Roadmap task: (3.14)

## Review walkthrough

- Start with [docs/netsuke-design.md](https://github.com/leynos/netsuke/blob/chore/conditional-action-deps/docs/netsuke-design.md) to review the design changes for action-level `foreach` / `when`, `command_available(...)`, manifest-time condition evaluation, dependency-field semantics, planned `deps_from`, and recipe ergonomics.
- Then review [docs/roadmap.md](https://github.com/leynos/netsuke/blob/chore/conditional-action-deps/docs/roadmap.md) for the new implementation tasks covering conditional actions, dependency lowering, command probes, `deps_from`, backend dollar escaping, Jinja helpers, structured `env`, structured `exec`, and selected-action feedback.
- Finish with [docs/users-guide.md](https://github.com/leynos/netsuke/blob/chore/conditional-action-deps/docs/users-guide.md) to confirm the unimplemented gcc-style dependency import capability has been removed from user-facing documentation.

## Validation

- `make markdownlint`: passed.
- `make nixie`: passed.
- `git diff --check`: passed.

## Notes

- `make fmt` was attempted during the documentation work, but it currently fails on pre-existing repository-wide Markdown line-length findings. Unrelated formatter churn was reverted, and the committed branch remains limited to the intended documentation files.
- This PR documents planned behaviour; it does not implement parser, IR, or Ninja backend changes.

## Summary by Sourcery

Document planned semantics and roadmap for conditional actions, dependency handling, and recipe ergonomics without changing implementation code.

Documentation:
- Clarify manifest-time behaviour of `foreach`/`when` for targets and actions, including how selected entries are lowered into the static IR and Ninja graph.
- Document the separation of `sources`, `deps`, and `order_only_deps`, and how they map to Ninja inputs and rebuild semantics.
- Specify planned rule-level `deps_from` for compiler dependency imports and remove the unimplemented `deps`-based example from the user guide.
- Describe planned backend dollar escaping, structured environment mappings via `env`, argv-safe `exec` recipes, and improved execution feedback via target/action descriptions.
- Add roadmap tasks covering conditional action support, dependency lowering, command availability probing, recipe environment/exec enhancements, and related testing.

### CODE REVIEW

File: docs/netsuke-design.md [hunk 1]
```diff
@@ -183,7 +183,10 @@ level keys.
   to a Ninja `build` statement.[^3]
 
 - `actions`: A secondary list of build targets. Any target placed here is
-  treated as `{ phony: true, always: false }` by default.
+  treated as `{ phony: true, always: false }` by default. Entries in this list
+  participate in the same manifest-time `foreach` and `when` expansion as
+  `targets`, so preparation and test actions can be selected declaratively
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/netsuke-design.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/netsuke-design.md (1)

`646-693`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Mark planned AST fields and recipe variants as planned, not implemented.**
Reframe this `src/ast.rs` block as planned schema, or align it with the current code.

`Rule.env`, `Target.description`, `Target.env`, and `Recipe::Exec` are shown as

current API here, but the referenced `src/ast.rs` surface does not expose them

yet.
As per coding guidelines, “When new decisions are made… update the relevant

file(s) in the `docs/` directory to reflect the latest state.”

##### 🤖 Prompt for AI Agents

File: docs/netsuke-design.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/netsuke-design.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/manifest/expand.rs (1)

`84-88`: _⚠️ Potential issue_ | _🟡 Minor_ | _⚡ Quick win_
**Reject whitespace-only `when` clauses.**
Harden Line 85 by trimming before the emptiness check; the current guard accepts whitespace-only input and silently evaluates it as false instead of invalid.

##### Patch

```diff
fn eval_when(env: &Environment, expr: &str, ctx: Value) -> Result<bool> {
     anyhow::ensure!(
-        !expr.is_empty(),
+        !expr.trim().is_empty(),
         "{}",
         localization::message(keys::MANIFEST_WHEN_EMPTY)
     );
```

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/netsuke-design.md (1)

`646-693`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Mark planned AST fields and recipe variants as planned, not implemented.**
Reframe this `src/ast.rs` block as planned schema, or align it with the current code.

`Rule.env`, `Target.description`, `Target.env`, and `Recipe::Exec` are shown as

current API here, but the referenced `src/ast.rs` surface does not expose them

yet.
As per coding guidelines, “When new decisions are made… update the relevant

file(s) in the `docs/` directory to reflect the latest state.”

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

🚥 Pre-merge checks | ✅ 16 | ❌ 2

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| User-Facing Documentation | ⚠️ Warning | PR adds foreach/when support for actions section, but docs/users-guide.md only documents this feature for targets, not actions. | Update section "### foreach and when" in docs/users-guide.md to clarify that these keys apply to both targets and actions, or add an example showing foreach/when with actions. |
| Developer Documentation | ⚠️ Warning | expand\_foreach now implements action expansion in code but design documentation still claims it is planned. No developer documentation added for this public API change to developers-guide.md. | Update docs/netsuke-design.md lines 411–413 and 1815 to reflect action-level foreach/when is implemented. Mark roadmap 3.14.2 complete if done. Document expand\_foreach in developers-guide.md. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 3 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests 1, 2, 4 lack explicit assertions checking foreach removal. Test 6 covers implicitly. Untested: empty foreach, non-object entries, variable overwrites, Jinja filters. | Add explicit foreach removal assertions in Tests 1, 2, 4. Add tests for empty foreach, non-object entries, variable overwrites, Jinja filters, non-iterable errors. |
| User-Facing Documentation | ⚠️ Warning | expand\_foreach now supports actions and targets, but docs/users-guide.md section on foreach/when only documents targets with targets-only example and prose. | Update docs/users-guide.md: change prose to state foreach/when apply to both targets and actions; add YAML example showing foreach/when under actions; update processing-order wording to mention actions. |
| Developer Documentation | ⚠️ Warning | Developer documentation incomplete: expand\_foreach not documented in developers-guide.md; roadmap 3.14.2 not marked complete; netsuke-design.md section 2.5 title inconsistent with implemented scope. | Add expand\_foreach documentation to developers-guide.md. Update roadmap 3.14.2 checkbox to \[x\]. Revise netsuke-design.md section 2.5 title to reflect targets and actions support. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces invariants over input ranges: foreach consistency, when filtering determinism, index sequentiality. No property-based testing to verify these across arbitrary inputs. | Add proptest property tests verifying: foreach expansion consistency for arbitrary item counts, when filtering determinism, index uniqueness/sequentiality, whitespace-handling uniformity. |
| Testing (Unit And Behavioural) | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ⚠️ Testing (Property / Proof) — Out of scope; GitHub issue proposed

Adding `proptest` requires a new dependency and a non-trivial amount of new code. This is out of scope for a documentation-and-expansion PR. Would you like me to open a GitHub issue to track adding proptest property tests for `expand_foreach`?

**Comment by leynos** [comment]
@coderabbitai this is out of scope for a documentation PR.

Please create a new GitHub issue for the following:

### ❌ Testing (Overall) — Not resolved

Tests 1, 2, and 4 have no explicit `foreach`\-removal assertion. Additionally, there is no coverage for empty `foreach`, non-object entries, variable overwrites, or Jinja filters.

**AI agent prompt:**

```
File to edit: src/manifest/expand_tests.rs

1. In `expand_foreach_expands_sequence_values` (Test 1), after confirming
   `targets.len() == 2`, add an assertion for each expanded target that its
   object map does not contain the key `"foreach"`.

2. In `expand_foreach_applies_when_expression` (Test 2), after confirming
   the filtered count, add an assertion for each remaining target that its
   object map does not contain the key `"foreach"`.

3. In `expand_foreach_applies_action_when_expression` (Test 4), after
   confirming the filtered count, add an assertion for each remaining action
   that its object map does not contain the key `"foreach"`.

4. Add a test `expand_foreach_empty_foreach_produces_no_entries` that
   passes a target with `foreach: []` and asserts the expanded `targets`
   slice is empty.

5. Add a test `expand_foreach_non_object_entry_is_passed_through` that
   passes a `targets` array containing a bare string value (non-object) and
   asserts it survives unexpanded.

6. Add a test `expand_foreach_iteration_vars_do_not_overwrite_globals` that
   passes a target whose `vars` map already contains a key `"item"` with a
   value, and asserts the iteration-injected `item` value (from `foreach`)
   takes precedence for each expanded entry (or whichever precedence rule
   `expand_section` implements — assert the actual observed behaviour).

7. Add a test `expand_foreach_jinja_filter_in_name` that passes a target
   whose `name` uses a Jinja filter (e.g. `{{ item | upper }}`) with a
   `foreach` list of lowercase strings and asserts the expanded names are
   uppercased correctly.

Do not remove or modify any existing tests.
```

File: docs/netsuke-design.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/netsuke-design.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The design now introduces several similarly named environment concepts (`env` mapping in manifests, `EnvValue`/`EnvOperation` in the AST, and `EnvBinding` in the IR); consider adding a short subsection explicitly describing how these layers relate and how conflicting operations (e.g., `value` + `prepend`) are resolved or rejected.
- The new `env()` Jinja helper signature and the manifest-level `env` mappings are easy to conflate; it would help to cross-reference them and state clearly that `env()` only reads from the process environment while the manifest `env` block only affects the execution environment Netsuke applies to actions.
- Several Rust-type snippets (e.g., `ExecRecipe`, `EnvValue`, `EnvBinding`) describe planned structures that are not yet implemented; consider adding a brief note in the design doc that these are forward-looking IR/API sketches so readers do not assume they already exist in the codebase.

## Individual Comments

### Comment 1
<location path="docs/netsuke-design.md" line_range="1650" />
<code_context>
     pub recipe: Recipe,
     pub description: Option<String>,
+    pub env: HashMap<String, EnvBinding>,
     pub depfile: Option<String>, // Template for the.d file path, e.g., "$out.d"
     pub deps_format: Option<String>, // "gcc" or "msvc"
     pub pool: Option<String>,
</code_context>
<issue_to_address>
**issue (typo):** Fix minor typo in `depfile` comment (`the.d` should have a space)

Update the inline comment to say `the .d file path` or `the ".d" file path` for clarity.

```suggestion
    pub depfile: Option<String>, // Template for the .d file path, e.g., "$out.d"
```
</issue_to_address>
~~~
